go/vt/vtctl/workflow/vexec/vexec.go

```diff
@@ -147,6 +154,52 @@ func (vx *VExec) QueryContext(ctx context.Context, query string) (map[*topo.Tabl
 	return qp.ExecuteScatter(ctx, vx.primaries...)
 }
 
+// CallbackContext executes the given callback, returning a mapping of tablet
+// to querypb.QueryResult.
+//
+// On first use, QueryContext will also cause the VExec instance to discover
+// target tablets from the topo; that target list will be reused for all future
+// queries made by this instance.
+//
+// For details on query parsing and planning, see GetPlanner and the
+// QueryPlanner interface.
+func (vx *VExec) CallbackContext(ctx context.Context, callback func(context.Context, *topo.TabletInfo) (*querypb.QueryResult, error)) (map[*topo.TabletInfo]*querypb.QueryResult, error) {
+	if vx.primaries == nil {
+		if err := vx.initialize(ctx); err != nil {
+			return nil, err
+		}
+	}
+	return vx.execCallback(ctx, callback)
+}
+
+// execCallback runs the provided callback function on backend shard primaries.
+// It collects query results from all shards and returns an aggregate (UNION
+// ALL -like) result.
+func (vx *VExec) execCallback(ctx context.Context, callback func(context.Context, *topo.TabletInfo) (*querypb.QueryResult, error)) (map[*topo.TabletInfo]*querypb.QueryResult, error) {
+	var wg sync.WaitGroup
+	allErrors := &concurrency.AllErrorRecorder{}
+	results := make(map[*topo.TabletInfo]*querypb.QueryResult)
+	var mu sync.Mutex
+	ctx, cancel := context.WithTimeout(ctx, execTimeout)
```

Comment:
nit: maybe condense these decls?
```suggestion
	var (
		wg sync.WaitGroup
		mu sync.Mutex
		
		allErrors = &concurrency.AllErrorRecorder{}
		results = make(map[*topo.TabletInfo]*querypb.QueryResult)
		ctx, cancel = context.WithTimeout(ctx, execTimeout)
	)
```